File: lintcode/unique-paths.py
#!/usr/bin/env python
# encoding: utf-8

# Question: http://www.lintcode.com/en/problem/unique-paths/

class Solution:
    """
    @param n and m: positive integer(1 <= n , m <= 100)
    @return an integer
    """
    def uniquePaths(self, m, n):
        # write your code here
        if m == 0 or n == 0:
            return 0

        matrix = [[0] * n for i in xrange(m)]
        for x in xrange(m):
            matrix[x][0] = 1
        for y in xrange(n):
            matrix[0][y] = 1

        for x in xrange(1, m):
            for y in xrange(1, n):
                matrix[x][y] = matrix[x-1][y] + matrix[x][y-1]

        return matrix[m-1][n-1]

if __name__ == '__main__':
    s = Solution()
    print s.uniquePaths(2, 3)

File: leetcode/unique-paths.cc
// Question: https://leetcode.com/problems/unique-paths/

#include "headers.h"

class Solution {
public:
    int uniquePaths(int m, int n) {
        vector<vector<int>> steps(m, vector<int>(n, 0));
        for (int i = 0; i < m; ++i) {
            steps[i][0] = 1;
        }
        for (int i = 0; i < n; ++i) {
            steps[0][i] = 1;
        }
        for (int i = 1; i < m; ++i) {
            for (int j = 1; j < n; ++j) {
                steps[i][j] = steps[i-1][j] + steps[i][j-1];
            }
        }
        return steps[m-1][n-1];
    }
};

int main(int argc, char** argv) {
    Solution s;
    cout << s.uniquePaths(1, 10) << endl;
}
